[Skia][GTK] Implement printing support

GeorgesStavracas opened the pull request:
#### 9046ee541dbffdf24a9e923afc0920cc4c910cbb
<pre>
[Skia][GTK] Implement printing support
<a href="https://bugs.webkit.org/show_bug.cgi?id=271153">https://bugs.webkit.org/show_bug.cgi?id=271153</a>

Reviewed by Carlos Garcia Campos.

Enable SkPDF in the build by adding the corresponding files, and their
dependencies, to the sources list in CMake.

Implement Skia code paths for WebPrintOperationGtk. Most of it is a
pretty direct conversion of cairo_t to SkCanvas APIs. The most notable
difference is that one SkPicture is created for each page sheet using
SkPictureRecorder, whereas Cairo operates on a single context for all
pages. When ending the print, take all SkPictures and add them in the
final document.

Limit WebKitPrintOperation to PDF files when printing using Skia, as
it&apos;s the only supported method for now. Modern printers support it, and
CUPS is able to implicitly convert between PDF and PostScript if that&apos;s
needed.

At last, change the Skia constructor of FragmentedSharedBuffer to receive
sk_sp&lt;SkData&gt;&amp;&amp; instead of a bare pointer (SkData*). This allows writing
the Skia variant of WebPrintOperationGtk::endPrint() in a more concise
way:

```
WebCore::FragmentedSharedBuffer::create(memoryBuffer.detachAsData())
```

instead of

```
sk_sp&lt;SkData&gt; data = memoryBuffer.detachAsData();
WebCore::FragmentedSharedBuffer::create(data.release());
```

* Source/ThirdParty/skia/CMakeLists.txt:
* Source/WebCore/platform/SharedBuffer.h:
* Source/WebCore/platform/skia/SharedBufferSkia.cpp:
(WebCore::FragmentedSharedBuffer::FragmentedSharedBuffer):
(WebCore::FragmentedSharedBuffer::create):
* Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
(webkitPrintOperationRunDialog):
* Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
(WebKit::WebPrintOperationGtk::startPrint):
(WebKit::WebPrintOperationGtk::startPage):
(WebKit::WebPrintOperationGtk::endPage):
(WebKit::WebPrintOperationGtk::endPrint):
(WebKit::WebPrintOperationGtk::rotatePageIfNeeded):
(WebKit::WebPrintOperationGtk::prepareContextToDraw):
(WebKit::WebPrintOperationGtk::renderPage):
(WebKit::WebPrintOperationGtk::printPagesDone):
(WebKit::WebPrintOperationGtk::print):
* Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.h:

Canonical link: <a href="https://commits.webkit.org/276477@main">https://commits.webkit.org/276477@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b0ac21e74640a7c2581e880c69337b561f37a15

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44762 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23856 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/47266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47416 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40767 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47065 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27904 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21249 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45340 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20932 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/47266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/17868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18357 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/47266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2809 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41006 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/47266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49077 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19727 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16298 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/21047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/47266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42525 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6196 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20721 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

GeorgesStavracas commented on Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
```diff
@@ -306,6 +325,63 @@ void WebPrintOperationGtk::endPrint(cairo_t* cr)
+
+    // FIXME: remove the extra copy
+    sk_sp<SkData> pdfData = m_skiaBuffer.detachAsData();
+    ASSERT(pdfData);
```
Thoughts?

carlosgcampos replied:
We could use `SharedBufferBuilder` like in the cairo case and add a custom SkStream that writes the data to the builder.

GeorgesStavracas replied:
Managed to replace this buffer copy with:

```cpp
sk_sp<SkData> pdfData = memoryBuffer.detachAsData();
WebCore::FragmentedSharedBuffer::create(pdfData.release());
```

GeorgesStavracas commented on Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
```diff
@@ -607,19 +737,31 @@ void WebPrintOperationGtk::printDone(RefPtr<WebCore::FragmentedSharedBuffer>&& b
+#if USE(CAIRO)
     m_cairoContext = adoptRef(cairo_create(surface));
+#elif USE(SKIA)
+    m_skiaDocument = document;
```
I suspect the memory management / refcounting is wrong here, but I can't prove how.

carlosgcampos replied:
I think it's ok, but this method could receive a `sk_sp<SkDocument>&&` instead and here you would `WTFMove(document)`

GeorgesStavracas commented on Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.h:
```diff
@@ -119,6 +129,10 @@ class WebPrintOperationGtk {
+#elif USE(SKIA)
+    SkDynamicMemoryWStream m_skiaBuffer;
+    sk_sp<SkDocument> m_skiaDocument;
+    SkCanvas* m_pageCanvas { nullptr };
```
I wonder if there's a more semantic way to store these fields. Perhaps with an auxiliary structure or class. Thoughts?

GeorgesStavracas reviewed (commented):
* Skia renders images with a seemingly lower scale, and they look blurry, I couldn't figure out why, all the code is scaling things properly.
* This branch triggers a crash in font handling code, and I couldn't yet track down what's going on there: https://gist.github.com/GeorgesStavracas/5da73fb5f5461aa57cdd03e8c732810b

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -63,16 +63,27 @@ GraphicsContextSkia::GraphicsContextSkia(sk_sp<SkSurface>&& surface, RenderingMo
 bool GraphicsContextSkia::hasPlatformContext() const
 {
-    return !!m_surface->getCanvas();
+    return !!m_surface->getCanvas() || m_canvas != nullptr;
```
We should null check m_surface here, when created with a canvas it will be nullptr.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -63,16 +63,27 @@ GraphicsContextSkia::GraphicsContextSkia(sk_sp<SkSurface>&& surface, RenderingMo
     RELEASE_ASSERT(m_surface->getCanvas());
 }
 
+GraphicsContextSkia::GraphicsContextSkia(SkCanvas* canvas, RenderingMode renderingMode, RenderingPurpose renderingPurpose)
```
I don't think we should pass mode and purpose, since it's always going to be the same, we can just initialize it.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -63,16 +63,27 @@ GraphicsContextSkia::GraphicsContextSkia(sk_sp<SkSurface>&& surface, RenderingMo
+    , m_renderingPurpose(renderingPurpose)
+{
+    RELEASE_ASSERT(m_canvas);
+}
```
What we actually need in both constructors is the canvas, but it's not refcounted, so if we get a canvas the caller needs to make sure the `GraphicsContext` is always going to be destroyed before the canvas. It's probably always the case, so we could have a single constructor receiving a `SkCanvas&` and stop passing the surface.  Maybe we can do this in a previous patch.

carlosgcampos replied:
Not always the case, in `ShareableBitmap` we create a `SkSurface` that we need to be owned by the graphics context. So, we need to keep the surface approach and add a constructor taking a `SkCanvas&`

carlosgcampos replied:
See https://github.com/WebKit/WebKit/pull/26185 what do you think?

GeorgesStavracas replied:
I like #26185, it's pretty much a better and more thorough version of what this patch does :slightly_smiling_face:

carlosgcampos commented on Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
```diff
@@ -261,12 +261,12 @@ static WebKitPrintOperationResponse webkitPrintOperationRunDialog(WebKitPrintOpe
-                                                                                                 | GTK_PRINT_CAPABILITY_SCALE));
+        | GTK_PRINT_CAPABILITY_NUMBER_UP_LAYOUT | GTK_PRINT_CAPABILITY_PAGE_SET | GTK_PRINT_CAPABILITY_REVERSE
+        | GTK_PRINT_CAPABILITY_COPIES | GTK_PRINT_CAPABILITY_COLLATE | GTK_PRINT_CAPABILITY_SCALE
+#if USE(SKIA)
+        | GTK_PRINT_CAPABILITY_GENERATE_PDF
+#endif
```
I'm surprised we don't need this in the cairo case too.

GeorgesStavracas replied:
Cairo has support for all three default formats (SVG, PDF, and PS).

This whole Unix printing API seems to be just a layer on Cairo, if you ask me. It's too Cairo-y not to be.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
```diff
@@ -50,6 +50,10 @@
+#elif USE(SKIA)
+#include <WebCore/GraphicsContextSkia.h>
+#include <skia/docs/SkPDFDocument.h>
+#include <skia/svg/SkSVGCanvas.h>
```
SVGCanvas?

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
```diff
@@ -232,6 +240,17 @@ void WebPrintOperationGtk::startPrint(WebCore::PrintContext* printContext, Compl
+    sk_sp<SkDocument> document;
+    if (!g_strcmp0(outputFormat, "pdf")) {
+        document = SkPDF::MakeDocument(&m_skiaBuffer);
+        ASSERT(document);
```
I've taken a look at the chromium implementation and they use `SkPictureRecorder`, so instead of writing directly to the PDF document, they record all pages into `SkPicture` and on document finish they create the PDF document and reply all the pages there with `SkCanvas::drawPicture()`. I think we could follow the same approach.

GeorgesStavracas replied:
Done that!

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
```diff
@@ -232,6 +240,17 @@ void WebPrintOperationGtk::startPrint(WebCore::PrintContext* printContext, Compl
+        notImplemented();
+        completionHandler(nullptr, WebCore::internalError(frameURL()));
+        return;
+    }
```
We should make sure the dialog doesn't show the options to print to PS and SVG

GeorgesStavracas replied:
That's what the `GTK_PRINT_CAPABILITY_GENERATE_PDF` flag in WebKitPrintOperation does :slightly_smiling_face: 

The documentation says:

> If neither GTK_PRINT_CAPABILITY_GENERATE_PDF nor GTK_PRINT_CAPABILITY_GENERATE_PS is specified, GTK assumes that all formats are supported.

And alas, when I added this, the three radio buttons in the print dialog disappeared.

carlosgcampos replied:
Ah! That's weird API, it's ok then.

carlosgcampos replied:
So, here we could probably assert if ps or svg or just ignore it, like any other value and always use pdf

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
```diff
@@ -306,6 +313,66 @@ void WebPrintOperationGtk::endPrint(cairo_t* cr)
+    }
+    document->close();
+
+    memoryBuffer.flush();
```
I think flush does nothing for memory streams

GeorgesStavracas replied:
Removed

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
```diff
@@ -306,6 +313,66 @@ void WebPrintOperationGtk::endPrint(cairo_t* cr)
+    sk_sp<SkData> pdfData = memoryBuffer.detachAsData();
+    ASSERT(pdfData);
+
+    printDone(WebCore::FragmentedSharedBuffer::create(pdfData.release()), { });
```
I don't know why I made  `FragmentedSharedBuffer::create()` receive a pointer, but it should receive a `sk_sp<SkData>&&` so you could just do 
```
printDone(WebCore::FragmentedSharedBuffer::create(memoryBuffer.detachAsData()), { });
```

GeorgesStavracas replied:
Changed to that!

GeorgesStavracas committed:
[Skia][GTK] Implement printing support
https://bugs.webkit.org/show_bug.cgi?id=271153

Reviewed by Carlos Garcia Campos.

Enable SkPDF in the build by adding the corresponding files, and their
dependencies, to the sources list in CMake.

Implement Skia code paths for WebPrintOperationGtk. Most of it is a
pretty direct conversion of cairo_t to SkCanvas APIs. The most notable
difference is that one SkPicture is created for each page sheet using
SkPictureRecorder, whereas Cairo operates on a single context for all
pages. When ending the print, take all SkPictures and add them in the
final document.

Limit WebKitPrintOperation to PDF files when printing using Skia, as
it's the only supported method for now. Modern printers support it, and
CUPS is able to implicitly convert between PDF and PostScript if that's
needed.

At last, change the Skia constructor of FragmentedSharedBuffer to receive
sk_sp<SkData>&& instead of a bare pointer (SkData*). This allows writing
the Skia variant of WebPrintOperationGtk::endPrint() in a more concise
way:

```
WebCore::FragmentedSharedBuffer::create(memoryBuffer.detachAsData())
```

instead of

```
sk_sp<SkData> data = memoryBuffer.detachAsData();
WebCore::FragmentedSharedBuffer::create(data.release());
```

* Source/ThirdParty/skia/CMakeLists.txt:
* Source/WebCore/platform/SharedBuffer.h:
* Source/WebCore/platform/skia/SharedBufferSkia.cpp:
(WebCore::FragmentedSharedBuffer::FragmentedSharedBuffer):
(WebCore::FragmentedSharedBuffer::create):
* Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
(webkitPrintOperationRunDialog):
* Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.cpp:
(WebKit::WebPrintOperationGtk::startPrint):
(WebKit::WebPrintOperationGtk::startPage):
(WebKit::WebPrintOperationGtk::endPage):
(WebKit::WebPrintOperationGtk::endPrint):
(WebKit::WebPrintOperationGtk::rotatePageIfNeeded):
(WebKit::WebPrintOperationGtk::prepareContextToDraw):
(WebKit::WebPrintOperationGtk::renderPage):
(WebKit::WebPrintOperationGtk::printPagesDone):
(WebKit::WebPrintOperationGtk::print):
* Source/WebKit/WebProcess/WebPage/gtk/WebPrintOperationGtk.h:

Canonical link: https://commits.webkit.org/276477@main
